tuf/ngclient/_internal/trusted_metadata_set.py

```diff
@@ -0,0 +1,464 @@
+                    self.timestamp.signed.version,
+                )
+            # Prevent rolling back snapshot version
+            if (
+                new_timestamp.signed.meta["snapshot.json"].version
+                < self.timestamp.signed.meta["snapshot.json"].version
+            ):
+                raise exceptions.ReplayedMetadataError(
+                    "snapshot",
+                    new_timestamp.signed.meta["snapshot.json"].version,
+                    self.timestamp.signed.meta["snapshot.json"].version,
+                )
+
+        if new_timestamp.signed.is_expired(self.reference_time):
+            raise exceptions.ExpiredMetadataError("New timestamp is expired")
+
+        self._trusted_set["timestamp"] = new_timestamp
+        logger.debug("Updated timestamp")
+
+    # TODO: remove pylint disable once the hash verification is in metadata.py
+    def update_snapshot(self, data: bytes):  # pylint: disable=too-many-branches
+        """Verifies and loads 'data' as new snapshot metadata.
+
+        Args:
+            data: unverified new snapshot metadata as bytes
+
+        Raises:
+            RepositoryError: Metadata failed to load or verify. The actual
+                error type and content will contain more details.
+        """
+
+        if self.timestamp is None:
+            raise RuntimeError("Cannot update snapshot before timestamp")
+        if self.targets is not None:
+            raise RuntimeError("Cannot update snapshot after targets")
+        logger.debug("Updating snapshot")
+
+        meta = self.timestamp.signed.meta["snapshot.json"]
+
+        # Verify against the hashes in timestamp, if any
+        hashes = meta.hashes or {}
+        for algo, stored_hash in hashes.items():
+            digest_object = sslib_hash.digest(algo)
+            digest_object.update(data)
+            observed_hash = digest_object.hexdigest()
+            if observed_hash != stored_hash:
+                raise exceptions.BadHashError(stored_hash, observed_hash)
+
+        try:
+            new_snapshot = Metadata.from_bytes(data)
+        except DeserializationError as e:
+            raise exceptions.RepositoryError("Failed to load snapshot") from e
+
+        if new_snapshot.signed.type != "snapshot":
+            raise exceptions.RepositoryError(
+                f"Expected 'snapshot', got '{new_snapshot.signed.type}'"
+            )
+
+        if not verify_with_threshold(self.root, "snapshot", new_snapshot):
+            raise exceptions.UnsignedMetadataError(
+                "New snapshot is not signed by root", new_snapshot.signed
+            )
+
+        if (
+            new_snapshot.signed.version
+            != self.timestamp.signed.meta["snapshot.json"].version
+        ):
+            raise exceptions.BadVersionNumberError(
+                f"Expected snapshot version"
+                f"{self.timestamp.signed.meta['snapshot.json'].version},"
+                f"got {new_snapshot.signed.version}"
```

Comment:
I think this is missing whitespace in the error message?
```suggestion
                f"Expected snapshot version "
                f"{self.timestamp.signed.meta['snapshot.json'].version}, "
                f"got {new_snapshot.signed.version}"
```